support/cas-server-support-pm-webflow/src/main/java/org/apereo/cas/pm/web/flow/PasswordManagementSingleSignOnParticipationStrategy.java

```diff
@@ -0,0 +1,26 @@
+package org.apereo.cas.pm.web.flow;
+
+import org.apereo.cas.web.flow.SingleSignOnParticipationStrategy;
+
+import lombok.val;
+import org.springframework.webflow.execution.RequestContext;
+
+/**
+ * This is {@link PasswordManagementSingleSignOnParticipationStrategy}.
+ *
+ * @author Julien Huon
+ * @since 6.3.0
+ */
+public class PasswordManagementSingleSignOnParticipationStrategy implements SingleSignOnParticipationStrategy {
+
+    @Override
+    public boolean supports(final RequestContext requestContext) {
+        val params = requestContext.getRequestParameters();
+        return params.contains(PasswordManagementWebflowUtils.REQUEST_PARAMETER_NAME_PASSWORD_RESET_TOKEN);
```

Comment:
I think this is a good starting point as a check. Is there something else, in addition to the request parameter check, that can be done to detect the request as a password-reset? Perhaps something in the RequestContext that track this execution as "this is a password reset"?

For example, a password-reset request should be:

- There is a token in the request
- It's valid token, has not been used, has not expired, etc.
- Only then, when all those checks pass should SSO be disabled.

I suspect those checks are already part of the flow, (you should not have to repeat them here) and their result is tracked in `RequestContext`, or should be if not already. 

There is no *danger* in this solution per se; I'd like to see if there is a way we could *harden* the check, and not allow bypassing of SSO only with a simple request parameter (that anyone any way can supply with any value).

Reply:
Thanks for the review.

There is an action wich is verifying if there is a token in the request and if it contains an username, it's called [InitPasswordResetAction](https://github.com/apereo/cas/blob/master/support/cas-server-support-pm-webflow/src/main/java/org/apereo/cas/pm/web/flow/actions/InitPasswordResetAction.java) and it's called in the pswdFlow subflow.

There is another one which is called just after the security question view: [verifyPasswordResetRequestAction](https://github.com/apereo/cas/blob/master/support/cas-server-support-pm-webflow/src/main/java/org/apereo/cas/pm/web/flow/actions/VerifyPasswordResetRequestAction.java) which is verifying if the token is valid.

I guess i could try to look in the requestContext if where in the view casMustChangePassView or VerifyPasswordResetRequestAction ?

Reply:
Yes, `VerifyPasswordResetRequestAction` is probably the best spot. It should be tracking the result of the token validation, of if not, you could add something to the request context that tracks that, and then use that in the strategy for disabling sso. 

The trick is to make sure that action and its result are executed and tracked before the sso strategy runs, so the result can be used. If they run after, then we can possibly add login in the strategy again to validate the token similar to how the action runs those validation rules.

Reply:
Understood. Will try it. :)

Reply:
The SSO Strategy run before the action. So, I've done what's you wanted in the strategy:
- Verify if there is a token in the request
- Verify if the token exists and if it's a TST

It's working if the token in the request is valid but when it's not, you're redirected to the service if you're using the cas protocol or you got the 500 error if you're using the oauth/oidc protocol...

So I guess it's not the best solution to add this logic here. Perhaps the pswrst parameter should be equivalent to a renew parameter if we want to fix the problem?

Reply:
- Both of those snippets do the same thing; perhaps one fails if the token is not found?

- You should not check if the token is a TST. You should just check that the token can be found, and can be parsed. Ignore its type.

- If it's not valid, then you should not disable SSO and should continue to support the request. But, if the token is not valid, some of the other actions that verify the token should fail. Failing is the correct behavior.

Reply:
Misagh,

The main point here is if I do not disable SSO when the ticket is not valid i'm redirected to the service (which is failing if it's an oauth/oidc one) because i'm logged in.

So if I do what you want i'm not fixing the problem in that kind of scenario. I'm only fixing it if the token is valid.

Reply:
- If you provide an invalid password reset token when you have SSO, do password-reset webflow actions run to validate the token at all?
- If you provide an invalid password reset token when you do not have SSO, do password-reset webflow actions run to validate the token at all?

Reply:
**If you provide an invalid password reset token when you have SSO, do password-reset webflow actions run to validate the token at all?**

No

**If you provide an invalid password reset token when you do not have SSO, do password-reset webflow actions run to validate the token at all?**

Yes

Reply:
OK, so that's where another problem lies. This is what should happen:

When a token is specified even with SSO, some webflow action must run to validate the token. If token is valid, the flow proceeds. If the token is invalid, then the flow should stop and must route to an error page, explaining the scenario. It sounds like this validation step is missing in the flow when there is SSO, so perhaps a new action needs to be added at the very beginning of the flow in the right stop to make sure if a valid token is specified for password resets.

Once you have this action in place, then the SSO strategy can do its job as it does now; validate the token and determine SSO etc.

I can take a look at the webflow validation action and try to fix it if it's easier for you to resume your work here.

Reply:
If you have SSO the request is see as a classic login request.

Reply:
I'll try to fix the webflow by myself first if you don't mind. :)

Reply:
Unless you want to fix this as soon as possible of course.